raiden/network/proxies/token_network.py

```diff
@@ -1341,78 +1393,166 @@ def update_transfer(
     def unlock(
         self,
         channel_identifier: ChannelID,
-        participant: Address,
-        partner: Address,
-        merkle_tree_leaves: Optional[MerkleTreeLeaves],
+        sender: Address,
+        receiver: Address,
+        merkle_tree_locks: MerkleTreeLeaves,
+        given_block_identifier: BlockSpecification,
     ):
         log_details = {
             "token_network": pex(self.address),
             "node": pex(self.node_address),
-            "partner": pex(partner),
-            "participant": pex(participant),
-            "merkle_tree_leaves": merkle_tree_leaves,
+            "sender": pex(sender),
+            "receiver": pex(receiver),
+            "merkle_tree_locks": merkle_tree_locks,
         }
 
-        if merkle_tree_leaves is None or not merkle_tree_leaves:
-            log.info("skipping unlock, tree is empty", **log_details)
-            return
+        if not merkle_tree_locks:
+            raise ValueError("unlock cannot be done without merkle_tree_leaves")
+
+        leaves = cast(List[Keccak256], [lock.lockhash for lock in merkle_tree_locks])
+        merkle_tree = MerkleTreeState(compute_layers(leaves))
+
+        # Check the preconditions for calling unlock at the time the event was
+        # emitted.
+        try:
+            channel_onchain_detail = self._detail_channel(
+                participant1=sender,
+                participant2=receiver,
+                block_identifier=given_block_identifier,
+                channel_identifier=channel_identifier,
+            )
+            sender_details = self._detail_participant(
+                channel_identifier=channel_identifier,
+                detail_for=sender,
+                partner=receiver,
+                block_identifier=given_block_identifier,
+            )
+        except ValueError:
+            # If `given_block_identifier` has been pruned the checks cannot be
+            # performed.
+            pass
```

Comment:
Looking at the PR, i am assuming that the main change revolves around here. In case we try to fetch the details of a channel, if the block identifier provided is pruned, then we try to unlock **anyway**? If that fails, and it fails because the target side of the channel was already unlocked, then we fail with a `RecoverableError`. That is, ignoring the fact that gas estimate failure and any other are unrelated to the block being pruned. 
Am i correct here?

Reply:
> then we try to unlock anyway? 

Yes, if preconditions cannot be checked because of a pruned block the proxy may send the transaction (depending on the estimate gas).

> If that fails, and it fails because the target side of the channel was already unlocked, then we fail with a RecoverableError

I don't understand this. What line of code are you talking about?